### ISSUE TITLE
[torch.compile] [Dynamic] SymInt doesn't work with FakeTensor.__contains__ 

### ISSUE THREAD

**Issue by shingjan** [pytorch/pytorch#98870]
### 🐛 Describe the bug

This Runtime error, `RuntimeError: Tensor.__contains__ only supports Tensor or scalar, but you passed in a <class 'torch.SymInt'>`, happened during compiling `EleutherAI/gpt-neo-125M` with `dynamic=True`. 

My minimal repro:
```
import torch

from torch._subclasses.fake_tensor import FakeTensorMode

a = 1
b = torch.ones([10])
print(a in b) # this works and evaluated to be True

with FakeTensorMode():
    a = torch.SymInt(1)
    b = torch.ones([10])
    a in b # failed
```

Stacktrace:
```
Traceback (most recent call last):
  File "test.py", line 12, in <module>
    a in b
  File "/home/yj/pytorch/torch/_tensor.py", line 1000, in __contains__
    raise RuntimeError(
RuntimeError: Tensor.__contains__ only supports Tensor or scalar, but you passed in a <class 'torch.SymInt'>.
```

However if one tweaks the above repro a little bit. Something like the following:
```
with FakeTensorMode():
    a = 1
    b = torch.ones([10])
    a in b # failed
```

The stacktrace will look very different with a different `DataDepedentOutputExcetpion`:
```
Traceback (most recent call last):
  File "test.py", line 12, in <module>
    a in b
  File "/home/yj/pytorch/torch/_tensor.py", line 998, in __contains__
    return (element == self).any().item()  # type: ignore[union-attr]
  File "/home/yj/pytorch/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "/home/yj/pytorch/torch/_subclasses/fake_tensor.py", line 1075, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
  File "/home/yj/pytorch/torch/_subclasses/fake_tensor.py", line 1257, in dispatch
    op_impl_out = op_impl(self, func, *args, **kwargs)
  File "/home/yj/pytorch/torch/_subclasses/fake_tensor.py", line 426, in local_scalar_dense
    raise DataDependentOutputException(func)
torch._subclasses.fake_tensor.DataDependentOutputException: aten._local_scalar_dense.default
```

Should a `__contains__` method be added to `FakeTensor`? Thanks!

cc: @ezyang @jansel @Chillee 

### Versions

PyTorch version: 2.1.0a0+git803a1a0
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.31

Python version: 3.8.15 (default, Nov 24 2022, 15:19:38)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-69-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 12.1.66
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3070
Nvidia driver version: 530.30.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.8.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.8.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.8.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.8.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.8.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.8.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.8.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
CPU(s):                          24
On-line CPU(s) list:             0-23
Thread(s) per core:              2
Core(s) per socket:              12
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       AuthenticAMD
CPU family:                      25
Model:                           33
Model name:                      AMD Ryzen 9 5900X 12-Core Processor
Stepping:                        0
Frequency boost:                 enabled
CPU MHz:                         2200.000
CPU max MHz:                     3700.0000
CPU min MHz:                     2200.0000
BogoMIPS:                        7400.26
Virtualization:                  AMD-V
L1d cache:                       384 KiB
L1i cache:                       384 KiB
L2 cache:                        6 MiB
L3 cache:                        64 MiB
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

**Comment by ngimel** [comment]
FakeTensor can't implement `__contains__` because that method requires peeking into actual contents of the tensor, and FakeTensor only has metadata.

**Comment by ezyang** [comment]
We can do the normal data dependent play book for this.

**Comment by wconstab** [comment]
@ezyang looks like you are working on this, feel free to dump it back in the oncall pile if not

**Comment by ezyang** [comment]
@shingjan do you have instructions for how to run the full gpt-neo with dynamic=True? Is it just standard method?

**Comment by shingjan** [comment]
@ezyang my repro looks like:
```
from transformers import GPTNeoForCausalLM, GPT2Tokenizer
import torch

model_id = "EleutherAI/gpt-neo-125M"
model = GPTNeoForCausalLM.from_pretrained(model_id)
tokenizer = GPT2Tokenizer.from_pretrained(model_id)

prompt = (
    "In a shocking finding, scientists discovered a herd of unicorns living in a remote, "
    "previously unexplored valley, in the Andes Mountains. Even more surprising to the "
    "researchers was the fact that the unicorns spoke perfect English."
)

input_ids = tokenizer(prompt, return_tensors="pt").input_ids


@torch._dynamo.optimize(dynamic=True)
def predict():
    gen_tokens = model.generate(
        input_ids,
        do_sample=True,
        temperature=0.9,
        max_length=100,
    )
    return tokenizer.batch_decode(gen_tokens)[0]

predict()
```

**Comment by ezyang** [comment]
The E2E probably still doesn't work

**Comment by shingjan** [comment]
@ezyang Thanks a lot for sending the fix. I tried e2e on gpt-neo-125m, got an error on `sympy.core.relational.Relational` not recognized as outputs in inductor.ir and came up with a [fix](https://github.com/pytorch/pytorch/pull/98971). I want to note that this pr is still a draft and may not be the final fix. Would be great if folks can take a look and see what can be improved or added as tests.

**Comment by shingjan** [comment]
closed with #98971
